lib/_emerge/depgraph.py

```diff
@@ -8177,6 +8177,7 @@ def _wrapped_select_pkg_highest_available_imp(
                     ):
                         return pkg, existing_node
 
+            matched_oldpkg = [x for x in matched_oldpkg if x in matched_packages]
             visible_matches = []
             if matched_oldpkg:
                 visible_matches = [
```

Comment:
Yes, the logic that follows is only valid when matched_oldpkg is a subset of matched_packages.